### ISSUE TITLE
Rank-3 residual stream with a unit dim is slower than the rank-2 equivalent

### ISSUE THREAD

**Issue by elogir** [tenstorrent/tt-mlir#8621]
## Summary

Two logically-equivalent transformer-shaped programs — same ops, same shardings, same number of layers — compile to substantially different TTNN IR and run at substantially different speeds depending only on whether the residual stream is shaped `[B*S, D]` (rank-2) or `[B, S=1, D]` (rank-3 with a unit dim).

## Reproducer

```python
import os, time
os.environ["CONVERT_SHLO_TO_SHARDY"] = "1"
import jax, jax.numpy as jnp, numpy as np
from jax.sharding import Mesh, NamedSharding, PartitionSpec

jax.config.update("jax_platforms", "tt")
B, S, D, N = 32, 1, 4096, 8
EPS = 1e-5
mesh = Mesh(np.array(jax.devices()), ("model",))

def rms(x, w):
    v = jnp.mean(jnp.square(x.astype(jnp.float32)), -1, keepdims=True)
    return (x * jax.lax.rsqrt(v + EPS).astype(x.dtype)) * w

def stack(x, ws):
    for wn, wl in ws:
        x = x + jnp.matmul(rms(x, wn), wl)
    return x

def run(name, x_shape, x_spec):
    x = jax.device_put(jnp.ones(x_shape, jnp.bfloat16),
                       NamedSharding(mesh, x_spec))
    ws = [(
        jax.device_put(jnp.ones((D,), jnp.bfloat16), NamedSharding(mesh, PartitionSpec(None))),
        jax.device_put(jnp.ones((D, D), jnp.bfloat16),
                       NamedSharding(mesh, PartitionSpec(None, "model"))),
    ) for _ in range(N)]
    c = jax.jit(stack, compiler_options={
        "export_path": "ir", "export_model_name": name}).lower(x, ws).compile()
    for _ in range(5): jax.block_until_ready(c(x, ws))
    t = time.perf_counter()
    for _ in range(50): out = c(x, ws)
    jax.block_until_ready(out)
    print(f"{name}: {(time.perf_counter()-t)/50*1000:.3f} ms")

with mesh:
    run("rank3_unit_S",  (B, S, D),  PartitionSpec(None, None, None))
    run("rank2_natural", (B*S, D),   PartitionSpec(None, None))
```

Here are the performance from this reproducer:

```
rank3_unit_S:  14.992 ms
rank2_natural:  5.529 ms
```

Note that I measured similar performance degradation on a full-sized LLM (Llama 3B/8B) with optimization level 1 or 2 and trace replay enabled.

## Possible resolution

Two logically-equivalent programs at different ranks should compile to nearly identical IR and run at nearly identical speeds. Frontends shouldn't need to flatten unit dims out of intermediates as a precondition for performance.

### I've not looked deeply into it but here is what an LLM said about it

The layout solver may be pruning sharded candidates more aggressively at rank-3 than at rank-2. `lib/Dialect/TTNN/Analysis/LegalTensorLayoutAnalysis.cpp:31-72` (`tensorShapeCompatibleWithShard`) rejects any sharded TILE-layout candidate where a tiled dim has fewer tiles than the corresponding grid dim — `tiledShape[i] < gridShape[i]` returns false. For rank-3 with `S=1`, that dim resolves to a single tile after tile-padding, disqualifying most sharded grids, so producer and consumer end up with disjoint candidate sets and the solver bridges them with extra reshape glue per layer. There is already a TODO at lines 45-48 from `rpavlovicTT` flagging the heuristic as worth revisiting via the op-model interface. Unverified — needs someone who knows the layout pass to confirm.

Tested with `pjrt-plugin-tt==1.2.0.dev20260526002843`
